src/extension/control-flow/src/main/java/org/geoserver/flow/ControlFlowCallback.java

```diff
@@ -187,51 +188,43 @@ public Operation operationDispatched(Request request, Operation operation) {
 
     public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
         this.applicationContext = applicationContext;
-    }
-
-    @Override
-    public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
-        if (GeoServerExtensions.bean(FlowControllerProvider.class) == null) {
-            beanFactory.initializeBean(provider, "defaultFlowControllerProvider");
+        if (applicationContext instanceof ConfigurableApplicationContext) {
+            // register default beans if needed
+            registDefaultBeansIfNeeded((ConfigurableApplicationContext) applicationContext);
+        } else {
+            // we cannot regist default beans, there is nothing else we can do about this
+            LOGGER.warning("Application context not configurable, control-flow default beans will not be registered.");
         }
-        
-        // look for a ControlFlowConfigurator in the application context, if none is found, use the
-        // default one
         provider = GeoServerExtensions.bean(FlowControllerProvider.class, applicationContext);
+        // default beans may have not been registered
         if (provider == null) {
             provider = new DefaultFlowControllerProvider(applicationContext);
         }
     }
 
-    @Override
-    public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry registry)
-            throws BeansException {
-        // Inject DefaultFlowControllerProvider definition it not defined
-        if (GeoServerExtensions.bean(FlowControllerProvider.class, applicationContext) == null) {
```

Comment:
Thanks for the feedback.

I'm not removing the use of ``GeoServerExtensions.bean(...)``, the beans are still retrieved by ``GeoServerExtensions.bean(...)`` see line 198.

However is TRUE that when I check if the beans exist in the application context I don't use the  ``GeoServerExtensions`` methods, long story short ``GeoServerExtensions`` caches the lookup result and will not update itself if the application context is updated (new beans added) or use the new updated context if provided:
https://github.com/geoserver/geoserver/blob/2.10.x/src/platform/src/main/java/org/geoserver/platform/GeoServerExtensions.java#L202-L202

so I used the application context to test if any bean exists, anyway since the find bean will no be used for anything except the existence test this should not be a problem.

Ideally the extension filter mechanism should be used, i.e. declare the default beans and if anyone wants to override them the extensions filter mechanism should be used. But since this was not used in the first lace adding it now will break backwards compatibility.